src/module_3/module_3.md

```diff
@@ -0,0 +1,891 @@
+      <td>0</td>
+      <td>2</td>
+      <td>33</td>
+      <td>42.0</td>
+      <td>31.134054</td>
+      <td>30</td>
+      <td>30.0</td>
+      <td>24.27618</td>
+    </tr>
+    <tr>
+      <th>2</th>
+      <td>33826472919172</td>
+      <td>ricepastapulses</td>
+      <td>3481384026244</td>
+      <td>4</td>
+      <td>0</td>
+      <td>0</td>
+      <td>0</td>
+      <td>0</td>
+      <td>0</td>
+      <td>0.081052</td>
+      <td>...</td>
+      <td>0</td>
+      <td>0</td>
+      <td>0</td>
+      <td>2</td>
+      <td>33</td>
+      <td>42.0</td>
+      <td>31.134054</td>
+      <td>30</td>
+      <td>30.0</td>
+      <td>24.27618</td>
+    </tr>
+    <tr>
+      <th>3</th>
+      <td>33826472919172</td>
+      <td>ricepastapulses</td>
+      <td>3291363377284</td>
+      <td>2</td>
+      <td>0</td>
+      <td>0</td>
+      <td>0</td>
+      <td>0</td>
+      <td>0</td>
+      <td>0.081052</td>
+      <td>...</td>
+      <td>0</td>
+      <td>0</td>
+      <td>0</td>
+      <td>2</td>
+      <td>33</td>
+      <td>42.0</td>
+      <td>31.134054</td>
+      <td>30</td>
+      <td>30.0</td>
+      <td>24.27618</td>
+    </tr>
+    <tr>
+      <th>5</th>
+      <td>33826472919172</td>
+      <td>ricepastapulses</td>
+      <td>3479090790532</td>
+      <td>3</td>
+      <td>0</td>
+      <td>0</td>
+      <td>0</td>
+      <td>0</td>
+      <td>0</td>
+      <td>0.081052</td>
+      <td>...</td>
+      <td>0</td>
+      <td>0</td>
+      <td>0</td>
+      <td>2</td>
+      <td>33</td>
+      <td>42.0</td>
+      <td>31.134054</td>
+      <td>30</td>
+      <td>30.0</td>
+      <td>24.27618</td>
+    </tr>
+  </tbody>
+</table>
+<p>5 rows × 24 columns</p>
+</div>
+
+
+
+### Training the model
+
+
+```python
+from sklearn.model_selection import train_test_split
+from sklearn.preprocessing import StandardScaler
+from sklearn.linear_model import LogisticRegression
+from sklearn.linear_model import SGDClassifier
+from sklearn.linear_model import RidgeClassifier
+```
+
+
+```python
+df['outcome'].value_counts()
+```
+
+
+
+
+    outcome
+    0    2132624
+    1      31329
+    Name: count, dtype: int64
+
+
+
+
+```python
+X = df.drop('outcome', axis=1)
+y = df['outcome']
+
+X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=0.4, random_state=9)
+```
+
+
+```python
+cat_cols = ['product_type', 'vendor']
+for c in cat_cols:
+    train_freq = X_train[c].value_counts()
+    X_train[c] = X_train[c].map(train_freq)
+    
+    test_freq = X_test[c].value_counts()
+    X_test[c] = X_test[c].map(test_freq)
+```
+
+
+```python
+sc = StandardScaler()
+
+X_train_scaled = sc.fit_transform(X_train)
+X_test_scaled = sc.transform(X_test)
+```
+
+
+```python
+from sklearn.model_selection import GridSearchCV
+from sklearn.metrics import classification_report
+from sklearn.model_selection import cross_validate
+from imblearn.over_sampling import SMOTE
+
+
+sgd_log = SGDClassifier(loss='log_loss',class_weight='balanced')
```

Comment:
I wouldn't use `balance` unless you have a very big class imbalance (which is not your case): it doesn't really change anything in terms of performance (try it yourself) and it distorts the probabilities produced by the model